Warehouse re-factor and enable/disable object Control

aeslaughter opened the pull request:
This is a re-organized and much improved version of the Warehouse re-factor to enable generic object enable/disable controls.

Let me explain a bit, there is too much, let me sum up:
1. The most critical portion of this change is the creation of `MooseObjectStorage`, which is a container for storing `MooseObject`s that automatically manages the "all", "boundary", "block" vectors of objects.
   (Note, there is also an `ObjectStorage` which I plan to use for non-MooseObject stuff, e.g. Actions).
2. On top of `MooseObjectStorage` there is an `ExecuteMooseObjectStorage` that simply contains a set of `MooseObjectStroage` containers, for each of the objects execution flag.
3. The `WarehouseBase` is completely abstract, in general the individual warehouses will still be highly custom. The warehouse then contain as many `MooseObjectStorage` containers as needed to accomplish the management of the various types (see AuxKernelWarehouse). These storage objects are then what is passed to the various computing threaded compute objects.

I am sure there will be some complaints, but be aware I went through about 5 iterations on this design. It was not easy to come up with something generic, but I really think that this is close.

aeslaughter committed:
Creates new Warehouse design and storage containers for objects

(refs #5680)

aeslaughter committed:
Update to MooseObject and Block/BoundaryRestrictable to work with new storage containers

This adds two helper classes that are virtually inherited; this allows the storage structure to store block/boundary restricted objects without performing a dynamic cast, it
is also necessary to avoid cyclic include statements with FEProblem when creating warehouses from WarehouseBase

(refs #5680)

aeslaughter committed:
DamperWarehouse: Update Damper objects to used new Warehouse

(refs #5680)

aeslaughter committed:
Add tests controlling active state of Dampers

(refs #5680)

aeslaughter committed:
ControlWarehouse: Convert Controls to a first-class MOOSE system

(refs #5680)

aeslaughter committed:
Create Control for enabling/disabling objects with time

(refs #5680)

aeslaughter committed:
AuxKernelWarhouse, AuxScalarKernelWarehouse: Convert AuxWarehouse into separate warehouse for AuxKernels and AuxScalarKerenls

(refs #5680)

aeslaughter committed:
Removes ControlInterface, which is not needed.

ControlInterface was added to allow Controls and the reporter postprocessor to share get methods, but
since the reporter only needs to read a parameter it doesn't need full access. Thus, the interface was removed
making the Control class complete and remove then chance of people trying to inherit from ControlInterface.

(refs #5680)

aeslaughter committed:
Tests the ability to enable/disable AuxKernel and AuxScalarKernel objects

This patch also generalizes the DisableObjects control to allow for multiple objects and
correspoinding start/end times.

(refs #5680)

permcody commented:
While I like the idea, I'm surprised that the execution of this cleanup added nearly 3x the code! I'll go through it and comment on it a bit, then we can discuss it further tomorrow.

permcody commented on framework/include/base/DependencyResolverInterface.h:
```diff
@@ -15,49 +15,100 @@
+
+template<typename T>
+void
+DependencyResolverInterface::cyclicDependencyError(CyclicDependencyException<T> & e, const std::string & header)
```
This is a nice improvement. Don't know why I never bothered to fact this out before.

permcody commented on framework/include/base/DependencyResolverInterface.h:
```diff
@@ -15,49 +15,100 @@
+
+template<typename T>
+void
+DependencyResolverInterface::sort(typename std::vector<T> & vector)
```
Are we only able to sort vectors? Why not keep the iterator pair instead?

aeslaughter replied:
Because of the vector MooseSharedPointers versus vector of pointers. When passing regular pointers an implicit cast occurs in the template methods, this is not possible with the shared pointers.

permcody commented on framework/include/base/MooseObject.h:
```diff
@@ -32,6 +82,8 @@ InputParameters validParams<MooseObject>();
  */
 class MooseObject :
+  public virtual BlockRestrictableHelper, // see BlockRestrictable.h
+  public virtual BoundaryRestrictableHelper, // see BoundaryRestrictable.h
```
We've done such a good job of keeping these out of the framework. You'll probably have to convince us that this is the best design.

friedmud replied:
I'm against this change.  It doesn't make sense at the most basic level: not all objects even act on geometry.  A list of objects that don't deal with geometry:

1. ScalarKernels
2. GeneralUserObjects
3. TimeIntegrators
4. TimeSteppers
5. Executioners
6. Problem

I'm sure there are others.

MooseObject is EVERYWHERE it must stay light.

In addition: absolutely 100% There should be no virtual inheritance in MOOSE

aeslaughter replied:
I knew this was going to be rejected, I don't like it much either. I will take another crack at fixing it: It was motivated by cyclic inheritance: FEProblem->NonlinearSystem->AuxKernelWarehouse->AuxKernel->BlockRestrictable->FEProblem

But, perhaps making the warehouses UniquePtr fixed this problem, I did that after I did this virtual inheritance stuff.

permcody replied:
You might sit down and actually draw a design (UML style) for this refactor. Basically you have two very powerful OO tools at your disposal: templating and polymorphism we are kind of halfway using both in this refactor, but I don't really feel that we've fixed any of the deficiencies of the current design. Yes I can see that the new organization does enable the control logic to work but it doesn't feel cleaner at all (i.e. we still have a unique warehouse for each object type with code in each one)

If you start from a base class and add a few intermediate classes for the differentness classes of objects the design should be abstract enough to have a reduced set of warehouses that can handle the cases where we have restrictions, execution types, and active lists. We need to fully enumerate all of these object classes. Also what are all of the ways that we need to reach into the warehouses to pull lists of objects out? Basically I want a defensible design that's is based around functional interfaces, not organization. It's not enough to say that if feels like we should split one big class into two because it's big. It should be split because you actually use the intermediate methods at some point in the framework. 

I'm not trying to be hard on you. I just think that a huge refactor of the core internals of MOOSE deserves a first class design.

aeslaughter commented:
There are a lot of changes because I added a lot of tests.

permcody commented on framework/include/base/MooseObject.h:
```diff
@@ -82,6 +152,9 @@ class MooseObject :
   const std::string & _name;
 
+  /// Reference to the "enable" InputParaemters, used by Controls for toggling on/off MooseObjects
+  const bool & _enabled;
```
While I like this idea. It does enable a whole new class of bugs. If objects can be turned on and off at will. The integrity of the system cannot be assured.

aeslaughter replied:
We could add an interface class for doing this: "Disablable" :) 

This way only objects that this makes sense for have the parameter.

permcody replied:
Perhaps, but do you trust the user to make this decision? If you disable a UO, Postprocessor, Kernel, BC, Material even an AuxKernel. You risk having the user make the mistake of not handling the "stale" values that would be received through the coupling. I can see a hundred ways for this to go bad but only very few ways for it to work properly. This might be a support nightmare.

aeslaughter replied:
On Fri, Nov 20, 2015 at 9:26 AM, Cody Permann <notifications@github.com>
wrote:

> In framework/include/base/MooseObject.h
> <https://github.com/idaholab/moose/pull/5985#discussion_r45487271>:
>
> > @@ -82,6 +152,9 @@ class MooseObject :
> >    /// The name of this object, reference to value stored in InputParameters
> >    const std::string & _name;
> >
> > +  /// Reference to the "enable" InputParaemters, used by Controls for toggling on/off MooseObjects
> > +  const bool & _enabled;
>
> Perhaps, but do you trust the user to make this decision?
>
Yes, many of them are quite capable of using this to solve more complex
problems.

> If you disable a UO, Postprocessor, Kernel, BC, Material even an
> AuxKernel. You risk having the user make the mistake of not handling the
> "stale" values that would be received through the coupling. I can see a
> hundred ways for this to go bad but only very few ways for it to work
> properly. This might be a support nightmare.
>
The only way to enable/disable an object based on time or the solve is by
creating a Control object, so if someone starts complaining about weired
behavior with old values and they have a Controls block, we will know what
to look for.

All this being said, I was under the impression from people wanting
controls (@andrsd) that this type of ability is the primary use. If this is
not the case then we can drop this PR and move on to something else.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5985/files#r45487271>.
>

aeslaughter replied:
Currently, we can enable/disable objects using time periods via TransientInterface (http://www.mooseframework.org/docs/doxygen/moose/classTransientInterface__inherit__graph_org.svg). What I am proposing here just makes this same idea a bit more general, but is essentially the same concept.

permcody commented on framework/include/base/MooseObjectStorage.h:
```diff
@@ -0,0 +1,398 @@
+template<typename MooseObjectType>
+void
+MooseObjectStorage<MooseObjectType>::updateActiveHelper(std::vector<MooseSharedPointer<MooseObjectType> > & active, const std::vector<MooseSharedPointer<MooseObjectType> > & all)
+{
```
This may be too expensive. We'll have to look at some non-trivial problems to understand. It may be better to cache this active lists.

aeslaughter replied:
This only happens once per timestep, on the updateActive calls.

aeslaughter replied:
Also we do loops all the time now, this PR already removes a few. Take a look at: `ComputeResidualThread::onBoundary`, which on each element calls:   `_sys.getBCWarehouse(_tid).activeIntegrated(bnd_id, bcs);`

```
void
BCWarehouse::activeIntegrated(BoundaryID boundary_id, std::vector<IntegratedBC *> & active_integrated) const
{
  active_integrated.clear();

  if (_bcs.find(boundary_id) != _bcs.end())
    for (std::vector<IntegratedBC *>::const_iterator it = _bcs.at(boundary_id).begin(); it != _bcs.at(boundary_id).end(); ++it)
      if ((*it)->isActive())
        active_integrated.push_back(*it);
}
```

Using `MooseObjectStorage` this vector would already be constructed and cached.

aeslaughter replied:
Holy Dooley: `isActive` is a beast as well to be calling on each element and side:
```
bool
TransientInterface::isActive()
{
  // if we have zero or one time periods -> all objects are active all the time
  if (_time_periods.size() <= 1)
    return true;

  // look if _t lies in one of our time periods
  for (unsigned int i=1; i <= _time_periods.size(); ++i)  // Careful! We are purposely indexing one past the end of the array
    if ((i == _time_periods.size()) || // Are we in the last time period?
        ((_time_periods[i-1]->start() <= _t) && (_t < _time_periods[i]->start())))  // OR are we in one of the intermediate periods?
    {
      bool ret_value;
      const std::vector<std::string> & objects = _time_periods[i-1]->getObjectList(_object_type, ret_value);
      if (std::find(objects.begin(), objects.end(), _ti_name) != objects.end())
        return ret_value;
      else
        return !ret_value;
    }

  return false;
}
```

friedmud replied:
Yep: this is an upgrade to many places in the code.

permcody commented on framework/include/base/MooseObjectStorage.h:
```diff
@@ -0,0 +1,398 @@
+  {
+    typename std::map<BoundaryID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator it;
+    for (it = _all_boundary_objects[tid].begin(); it != _all_boundary_objects[tid].end(); ++it)
+      updateActiveHelper(_active_boundary_objects[tid][it->first], it->second);
```
I notice this pattern of having to deal with block and boundary objects all the time. They have all the same methods. We should think about whether or not it's possible to have a single restricted class. Yes I know that you have this dual-restrictable thingy... Who uses that? Is there another way?

aeslaughter replied:
Only a few classes can be both ("dual_restrictable"), most can be one or the other. I am for re-factoring this, but this is another issue: https://github.com/idaholab/moose/issues/4214

friedmud replied:
I'm ok with having them separate.  It's super clear this way... because, like @aeslaughter said, many classes are only one or the other.

permcody commented on framework/include/base/ObjectStorage.h:
```diff
@@ -0,0 +1,106 @@
+ * A general storage container for objects capable of storing threaded copies.
+ */
+template<typename ObjectType>
+class ObjectStorage
```
I'm having a difficult time seeing the need for ObjectStorage and MooseObjectWarehouse

permcody replied:
oh and WarehouseBase too.

permcody replied:
and ExecuteObjectStorage...

aeslaughter replied:
Please see the description in the PR behind the reason of `ObjectStorage`, I haven't used it for its purpose at this point, but will.

`MooseObjectWarehouse` is simply a convenience class for storage of single object types: Dampers, Controls, ...

All warehouses that have multiple objects (AuxKernels, UserObjects, ...) will inherit from WarehouseBase and use a custom set of storage containers, see AuxKernelWarehouse

permcody commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -137,17 +137,11 @@ InitialConditionWarehouse::sortICs(std::vector<InitialCondition *> & ics)
-    for (std::multimap<DependencyResolverInterface *, DependencyResolverInterface *>::const_iterator it = depends.begin(); it != depends.end(); ++it)
-      oss << (static_cast<InitialCondition *>(it->first))->name() << " -> " << (static_cast<InitialCondition *>(it->second))->name() << std::endl;
-    mooseError(oss.str());
+    DependencyResolverInterface::cyclicDependencyError<InitialCondition *>(e, "Cyclic dependency detected in InitialCondition ordering");
```
Nice!

permcody commented on framework/src/postprocessors/RealControlParameterReporter.C:
```diff
@@ -12,30 +12,31 @@
 {
-  _parameter = &getControllableValue<Real>("parameter");
+  MooseObjectParameterName name(getParam<std::string>("parameter"));
+  const InputParameters & params = getMooseApp().getInputParameterWarehouse().getInputParametersObject(name.tag(), name.name(), _tid);
```
Wait! Would a user write a line of code like this?

aeslaughter replied:
No. This is just a tool for viewing a parameter value, I could move this to moose-test

aeslaughter committed:
Implements organization changes based on developer comments.

- Removed virtual inheritence crazyness
- Separates Nodal and Elemental AuxKernels into different warehouses
- MooseObjectWarehouse is now the base class; WarehouseBase removed
- ObjectStorage removed and implemented in MooseObjectStorage
- Moves RealControlParameterPostprocessor to moose-test

(refs #5680)

aeslaughter commented:
I removed the virtual garbage, don't know why I didn't see the fix when I did this originally.

friedmud commented on framework/include/base/MooseObjectStorage.h:
```diff
@@ -0,0 +1,408 @@
+  bool has_active_block_objects = false;
+  typename std::map<SubdomainID, std::vector<MooseSharedPointer<MooseObjectType> > >::const_iterator it;
+  for (it = _active_block_objects[tid].begin(); it != _active_block_objects[tid].end(); ++it)
+    has_active_block_objects |= !(it->second.empty());
```
Might as well short circuit once you find one.  No reason to keep iterating...

friedmud commented:
How does this new stuff handle objects that are NOT threaded?  Does it finally do away with threaded copies of GeneralUserObjects?

aeslaughter commented:
Yes, this would do away threaded copies, `MooseObjectStorage` takes a single argument ("threaded"), when set to "false" only a single copy is stored.

aeslaughter commented:
The following decisions were made:
1. `MooseObjectWarehouse` should be the base class and Elemental and Nodal AuxKerenls (an other similar system) should be separated into multiple warhouses.
2. ActionWarehouse and other non-MooseObject stuff should just remain a unique warehouse design.
3. A method for not allowing enable/disable of objects should exist for Executioner, FEProblem, etc.
4. There should be a means to display what is enabled/disabled as the solve progress (another issue)
![img_0054](https://cloud.githubusercontent.com/assets/341333/11311232/bc5e2464-8f8b-11e5-92b8-4105cab593d4.jpg)

aeslaughter committed:
Objects must define if the 'enable' parameter is controllable

This allows us to limit the ability to control objects such as FEProblem, etc.

(refs #5680)

aeslaughter commented:
I have a PR for YAK waiting.
